src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.Worker.cs

```diff
@@ -0,0 +1,499 @@
+                        {
+                            continue; // Failure should already be reported.
+                        }
+
+                        if (!processedPaths.Add(absoluteProjectPath))
+                        {
+                            // TODO: Report warning if there are duplicate project paths.
+                            continue;
+                        }
+
+                        var projectFileInfos = await LoadProjectInfosFromPathAsync(absoluteProjectPath, _requestedProjectOptions, cancellationToken).ConfigureAwait(false);
+
+                        results.AddRange(projectFileInfos);
+                    }
+
+                    foreach (var (projectPath, projectInfos) in _pathToDiscoveredProjectInfosMap)
+                    {
+                        if (!processedPaths.Contains(projectPath))
+                        {
+                            results.AddRange(projectInfos);
+                        }
+                    }
+
+                    return results.ToImmutable();
+                }
+                finally
+                {
+                    _buildManager.Stop();
+                }
+            }
+
+            private async Task<ImmutableArray<ProjectFileInfo>> LoadProjectFileInfosAsync(string projectPath, DiagnosticReportingOptions reportingOptions, CancellationToken cancellationToken)
+            {
+                if (!_projectFileLoaderRegistry.TryGetLoaderFromProjectPath(projectPath, reportingOptions.OnLoaderFailure, out var loader))
+                {
+                    return ImmutableArray<ProjectFileInfo>.Empty; // Failure should already be reported.
+                }
+
+                var projectFile = await DoOperationAndReportProgressAsync(
+                    ProjectLoadOperation.Evaluate,
+                    projectPath,
+                    targetFramework: null,
+                    () => loader.LoadProjectFileAsync(projectPath, _globalProperties, _buildManager, cancellationToken)
+                ).ConfigureAwait(false);
+
+                // If there were any failures during load, we won't be able to build the project. So, bail early with an empty project.
+                if (projectFile.Log.HasFailure)
+                {
+                    _diagnosticReporter.Report(projectFile.Log);
+
+                    return ImmutableArray.Create(
+                        ProjectFileInfo.CreateEmpty(loader.Language, projectPath, projectFile.Log));
+                }
+
+                var projectFileInfos = await DoOperationAndReportProgressAsync(
+                    ProjectLoadOperation.Build,
+                    projectPath,
+                    targetFramework: null,
+                    () => projectFile.GetProjectFileInfosAsync(cancellationToken)
+                ).ConfigureAwait(false);
+
+                var results = ImmutableArray.CreateBuilder<ProjectFileInfo>();
```

Comment:
>() [](start = 75, length = 2)

Can specify `projectFileInfos.Count` as the builder length here.

Reply:
Sure. What is `MoveToImmutable`?

Reply:
Oh, [I see](http://source.roslyn.io/#System.Collections.Immutable/System/Collections/Immutable/ImmutableArray_1.Builder.cs,36cca4495ee0d898). Sure.